src/cls/cmpxattr/server.cc

```diff
@@ -0,0 +1,169 @@
+      uint64_t lhs;
+      auto p = input.cbegin();
+      using ceph::decode;
+      decode(lhs, p);
+      return compare_values_u64(op, lhs, value);
+    } catch (const buffer::error&) {
+      // failures to decode input values are reported as EINVAL
+      return -EINVAL;
+    }
+  default:
+    return -EINVAL;
+  }
+}
+
+static int cmp_vals_set_vals(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  cmp_vals_set_vals_op op;
+  try {
+    auto p = in->cbegin();
+    decode(op, p);
+  } catch (const buffer::error&) {
+    CLS_LOG(0, "ERROR: %s: failed to decode input", __func__);
+    return -EINVAL;
+  }
+
+  // read the values for each key to compare
+  std::map<std::string, bufferlist> cmp_values;
+  for (const auto& kv : op.cmp_pairs) {
+    const std::string &key = kv.first;
+    ceph::bufferlist bl;
+    int ret = cls_cxx_getxattr(hctx, key.c_str(), &bl);
+    if (ret < 0 && ret != -ENODATA) {
+      CLS_LOG(4, "ERROR: %s: cls_cxx_getxattr() for key=%s ret=%d",
+	      __func__, key.c_str(), ret);
+      return ret;
+    } else {
+      // ENODATA will generate an empty value bl
+      cmp_values[key] = bl;
+    }
+  }
+
+  auto v = cmp_values.begin();
+  for (const auto& [key, input] : op.cmp_pairs) {
+    auto k = cmp_values.end();
+    bufferlist value;
+    if (v != cmp_values.end() && v->first == key) {
+      value = std::move(v->second);
+      k = v++;
+      CLS_LOG(20, "%s::comparing key=%s mode=%d op=%d",
+	      __func__, key.c_str(), (int)op.mode, (int)op.comparison);
+    } else {
+      CLS_LOG(10, "%s:: missing key=%s, abort operation", __func__, key.c_str());
+      return -EINVAL;
+    }
+
+    // an empty input value will match an empty value or a non-existing key
+    if (input.length() == 0 && value.length() > 0) {
+      CLS_LOG(1, "%s:: key=%s exists!", __func__, key.c_str());
+      return -EEXIST;
+    }
+    int ret = compare_value(op.mode, op.comparison, input, value);
+    if (ret <= 0) {
+      // unsuccessful comparison
+      CLS_LOG(10, "%s:: failed compare key=%s ret=%d", __func__, key.c_str(), ret);
+      if (ret < 0) {
+	return ret;
+      }
+      else {
+	// ret value of 0 is understood as false here for compare failure
+	// TBD: return a better error code to be able to distinguish from a
+	//      generic failure
+	return -1;
+      }
+    }
+
+    // successful comparison
+    CLS_LOG(20, "%s:: successful comparison key=%s", __func__, key.c_str());
+  }
+
+  // if arrived here all keys in the cmp_pairs passed check
+  // overwrite all key/values in the set_pairs
+  for (const auto& [key, value] : op.set_pairs) {
+    auto inbl = const_cast<ceph::bufferlist *>(&value);
+    int ret = cls_cxx_setxattr(hctx, key.c_str(), inbl);
+    if (ret == 0) {
+      CLS_LOG(20, "%s:: successful set xattr key=%s", __func__, key.c_str());
+    }
+    else {
+      CLS_LOG(4, "ERROR: %s failed to set xattr key=%s ret=%d", __func__, key.c_str(), ret);
+      return ret;
```

Comment:
isn't it better to continue until the end and skip the failed sets? or maybe rollback in case of an error?

Reply:
It is all or nothing operation, even if a single cmp failed the whole operation will be aborted